### DISCUSSION TITLE
How to rotate ca cert with new root and --force

### DISCUSSION THREAD

**Discussion by nonessentialprototype** [rancher/rke2#5613] [Q&A] [answered]
I am on rke2 2.24.17, and I am trying to rotate the ca cert with a newly generated ca certs.  We want to invalidate the old certs/kube config so we do NOT want it cross signed.

I've followed the step here modified for my needs and am running this on the first server
https://docs.rke2.io/security/certificates#rotating-custom-ca-certificates
```
mkdir -p /tmp/rke2/server/tls
cp  /var/lib/rancher/rke2/server/tls/service.key /tmp/rke2/server/tls

DATA_DIR=/tmp/rke2 PRODUCT=rke2 ./generate-custom-ca-certs.sh

rke2 certificate rotate-ca --path=/tmp/rke2/server --server=https://127.0.0.1:9345 --force
```
The rotate is successfully but after I run `systemctl restart rke2-server.service` the server never comes back up and and logs show 

`CA cert validation failed: Get "https://127.0.0.1:9345/cacerts" tls: failed to verify certificate x509 certificate signed by unknown authority.`

This section also mentions
```
If you used the --force option or changed the root CA, ensure that any nodes that were joined with a secure token are reconfigured to use the new token value, prior to being restarted. The token may be stored in a .env file, systemd unit, or config.yaml, depending on how the node was configured during initial installation.
```
which seems to be a copy paste error cause the generate-custom-ca-certs does not output a new token.

I then tried these steps modifying the rotate script to not copy and use the old ones.
https://docs.rke2.io/security/certificates#rotating-self-signed-ca-certificates
with the same result, it rotates successfully but 509s on restart.

**Comment by brandond** [comment, marked as answer]
> generate-custom-ca-certs does not output a new token.

yes, generate-custom-ca-certs was intended to be used before initial cluster startup. Using that that to switch from the default self-signed certs, to custom certs, is not something that we've tested. You'll need to calculate the hash yourself, and update the hash in the token files by hand.

**Reply by nonessentialprototype** [reply]
If that is the case the documentation here https://docs.rke2.io/security/certificates#using-the-example-script-1
, and the script should be updated as it even prints out
```
  echo "To update certificates on an existing cluster, you may now run:"
  echo "    k3s certificate rotate-ca --path=${DATA_DIR}/server"
 ```

**Reply by brandond** [reply]
Yes, it prints that because it is intended to also be used to update the custom certs when they need to be renewed. It is just use of that script to switch from default to custom that isn't a tested path.

**Comment by nonessentialprototype** [comment]
I have been able to rotate successfully once, subsequent attempts after all the nodes are up fail. 

I ran across this comment https://github.com/k3s-io/k3s/issues/8952#issuecomment-1828689525

Which got me thinking to modify the `rotate-default-ca-certs.sh`

Adding an If around ${TYPE}-ca.crt creation

```
if [[ $TYPE = "client" ]]; then
  cat ${TYPE}-ca.pem \
  ${TYPE}-root-ssigned.pem > ${TYPE}-ca.crt

else
    cat ${TYPE}-ca.pem \
      ${TYPE}-root-ssigned.pem \
      ${TYPE}-root-xsigned.pem \
      ${TYPE}-root-old.pem > ${TYPE}-ca.crt
fi
```
 
then running 

`PRODUCT=rke2 ./rotate-default-ca-certs.sh`
`rke2 certificate rotate-ca --path=/var/lib/rancher/rke2/server/rotate-ca --server=https://127.0.0.1:9345 --force`
`rke2 certificate rotate`
`systemctl restart rke2-server`

once up, updating the config on each server/agent to the new token and then rebooting them.

Once they are all rebooted, the original kubeconfig has become invalidated.

**Reply by hdong69** [reply]
hi @nonessentialprototype,

Could you please explain more about the step '**updating the config on each server/agent to the new token and then rebooting them**'. What actually i have to do with the token and why ?

**Reply by nonessentialprototype** [reply]
It was the 'etc/rancher/rke2/config.yaml' on each agent and node with the new token generated above.

I dont have access currently but we were able to rotate to successfully multiple times in a row. 

Either way it's more of a pain then it should be and rke2 should really not use the cert based kubeconfig as the default

**Reply by brandond** [reply]
> @nonessentialprototype
> rke2 should really not use the cert based kubeconfig

... how exactly would you suggest we handle authentication for core Kubernetes components without client certificates?

**Reply by nonessentialprototype** [reply]
I think for the initial setup using the cert is fine, but as soon as one node is up a service account should be added and used.  This rbac admin "rke2" user would be what is exposed in the rke2.yaml and can then be revoked at any time.

Can you still build the cert based rbac sure, but that at least requires the access to those nodes and files.  

If the cert based approach is the ideal one then at least provide a script to easily revoke and rotate them when needed.

**Reply by brandond** [reply]
The issue is that golang does not check CRLs. So revoking certificates is a waste of time.

I'm not aware of any Kubernetes distros that use service account tokens for the admin or core component kubeconfig. IIRC it's really not practical as tokens are short lived and need to be validated against the oidc endpoint.

**Reply by nonessentialprototype** [reply]
I think we are looking at different views. Rke2 doing  cluster things if it needs to be cert based that's fine, should it be easier to force a cert rotation and invalidate old ones that's debatable.  However from a user standpoint. That rke2.yaml kubeconfig on each control plane shouldn't be that cert based config and should be a long running service account backed rbac. If whomever sets up that cluster exposes that file it shouldn't then become a pain to invalidate.

**Reply by brandond** [reply]
If that file is exposed your cluster is compromised, even if you rotate certs. That kubeconfig has full admin access and may have already pivoted to nodes or other accessible infrastructure. The threat model of compromised access to the admin kubeconfig does not change by eliminating client certs or making it easier to revoke existing certs.

**Reply by nonessentialprototype** [reply]
No but there is a way to do a rotation and force a new cert thus invalidating the exiting cert based rbac it's not straightforward.  My whole point is the kubeconfig rke2 exposes to the end user after a cluster is set up shouldn't be the cert one.

**Reply by hdong69** [reply]
> It was the 'etc/rancher/rke2/config.yaml' on each agent and node with the new token generated above.
> 
> I dont have access currently but we were able to rotate to successfully multiple times in a row.
> 
> Either way it's more of a pain then it should be and rke2 should really not use the cert based kubeconfig as the default

@nonessentialprototype 
1. For the image below, these red blocks should be updated in the config.yaml, right ?
![image](https://github.com/user-attachments/assets/7e57b6aa-b308-406b-aa7b-8ce99782b1f9)

2.
`PRODUCT=rke2 ./rotate-default-ca-certs.sh
rke2 certificate rotate-ca --path=/var/lib/rancher/rke2/server/rotate-ca --server=https://127.0.0.1:9345 --force
rke2 certificate rotate
systemctl restart rke2-server`

For your script you provided, do i need to run it on all master nodes ?

**Reply by nonessentialprototype** [reply]
Yes for 1
I believe it was just on the first node then would you update the masters configs and  restart those.

**Reply by hdong69** [reply]
many thanks @nonessentialprototype
